### ISSUE TITLE
[Proposal] Add Nats Streaming Parallel Concurrency Mode

### ISSUE THREAD

**Issue by Admiralkheir** [dapr/components-contrib#1669]
## Expected Behavior
As in version 1.0.1, sidecar should sends the events without any limitation on its side, for the previous event It should forward the event without waiting for a success or fail response. We make this limitation with the maxinflight parameter on the nats streaming side. If Nats does not receive an ack, it transmits the event itself.

## Actual Behavior and Problem
Hi, We were using nats streaming as a service broker with dapr version 1.0.1. Two weeks ago we updated dapr to the 1.6.1 version and noticed pub/sub behaviors changed. We noticed sidecar is sending events to app one by one. Sending one and waiting for a response (success or failure) if success it is sending a second event but nats streaming component has max inflight parameter for this (unack message count). We are setting this parameter 5 and noticed at subscription log, nats is sending 5 messages to sidecar but sidecar is sending one by one and waiting for processing event. This behaviour impact concurrency I think. 

Another problem is retry mechanism, nats streaming has retry mechanism natively. is it possible to close the retry mechanism on the subscription side? I tried Resiliency options and it does not affect. Sidecar keeps retrying for failed event forever

component.yaml
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: messagebus
spec:
  metadata:
    - name: natsURL
      value: >-
        localhost:4222
    - name: natsStreamingClusterID
      value: test-cluster
    - name: subscriptionType
      value: topic
    - name: deliverAll
      value: true
    - name: ackWaitTime
      value: 45000ms
    - name: maxInFlight
      value: '5'
  type: pubsub.natsstreaming
  version: v1
```
resiliency.yaml
```yaml
apiVersion: dapr.io/v1alpha1
kind: Resiliency
metadata:
  name: myresiliency
# similar to subscription and configuration specs, scopes lists the Dapr App IDs that this
# resiliency spec can be used by.
scopes:
  - app
  - app2
spec:
  # policies is where timeouts, retries and circuit breaker policies are defined. 
  # each is given a name so they can be referred to from the targets section in the resiliency spec.
  policies:
    # timeouts are simple named durations.
    timeouts:
      general: 5s
      important: 60s
      largeResponse: 10s

    # retries are named templates for retry configurations and are instantiated for life of the operation.
    retries:
      pubsubRetry:
        policy: constant
        duration: 10s
        maxRetries: 1

      retryForever:
        policy: exponential
        maxInterval: 15s
        maxRetries: -1 # retry indefinitely

    # circuit breakers are automatically instantiated per component and app instance.
    # circuit breakers maintain counters that live as long as the Dapr sidecar is running. They are not persisted.
    circuitBreakers:
      simpleCB:
        maxRequests: 1
        timeout: 30s 
        trip: consecutiveFailures >= 5

      pubsubCB:
        maxRequests: 1
        interval: 0s
        timeout: 45s
        trip: consecutiveFailures > 3

  # targets are what named policies are applied to. Dapr supports 3 target types - apps, components and actors
  targets:
    apps:
      app:
        retry: pubsubRetry

    components:
      # for state stores, policies apply to saving and retrieving state.
      messagebus: # any component name -- happens to be a pubsub broker here
        outbound:
          retry: pubsubRetry
        inbound: # inbound only applies to delivery from sidecar to app
          # timeout: general
          retry: pubsubRetry
```
dapr's log:
```log
time="2022-04-13T17:59:51.4039915+03:00" level=info msg="starting Dapr Runtime -- version 1.7.0 -- commit 3c4510a685470eb8a5403e9d9a451bc93e67e6c4" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.4039915+03:00" level=info msg="log level set to: debug" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.4039915+03:00" level=info msg="metrics server started on :9090/" app_id=app instance=PF1SE31X scope=dapr.metrics type=log ver=1.7.0
time="2022-04-13T17:59:51.4222667+03:00" level=debug msg="Found 2 resiliency configurations." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.4497547+03:00" level=info msg="standalone mode configured" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.4497547+03:00" level=info msg="app id: app" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.4497547+03:00" level=info msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5046499+03:00" level=info msg="local service entry announced: app -> 192.168.1.41:56653" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T17:59:51.5046499+03:00" level=info msg="Initialized name resolution to mdns" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5051636+03:00" level=info msg="loading components" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5108289+03:00" level=debug msg="found component. name: messagebus, type: pubsub.natsstreaming/v1" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5108289+03:00" level=info msg="waiting for all outstanding components to be processed" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5108289+03:00" level=debug msg="loading component. name: messagebus, type: pubsub.natsstreaming/v1" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5232512+03:00" level=debug msg="connected to natsstreaming at localhost:4222" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T17:59:51.5236309+03:00" level=info msg="component loaded. name: messagebus, type: pubsub.natsstreaming/v1" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5236309+03:00" level=info msg="all outstanding components processed" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5236309+03:00" level=info msg="gRPC proxy enabled" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5241076+03:00" level=info msg="enabled gRPC tracing middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.grpc.api type=log ver=1.7.0
time="2022-04-13T17:59:51.5241076+03:00" level=info msg="enabled gRPC metrics middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.grpc.api type=log ver=1.7.0
time="2022-04-13T17:59:51.5241076+03:00" level=info msg="API gRPC server is running on port 50001" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5244036+03:00" level=info msg="enabled metrics http middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.http type=log ver=1.7.0
time="2022-04-13T17:59:51.5244036+03:00" level=info msg="enabled tracing http middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.http type=log ver=1.7.0
time="2022-04-13T17:59:51.5252444+03:00" level=info msg="http server is running on port 3500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5252444+03:00" level=info msg="The request body size parameter is: 4" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5256154+03:00" level=info msg="enabled gRPC tracing middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.grpc.internal type=log ver=1.7.0
time="2022-04-13T17:59:51.5257206+03:00" level=info msg="enabled gRPC metrics middleware" app_id=app instance=PF1SE31X scope=dapr.runtime.grpc.internal type=log ver=1.7.0
time="2022-04-13T17:59:51.5257206+03:00" level=info msg="internal gRPC server is running on port 56653" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:51.5257206+03:00" level=info msg="application protocol: http. waiting on port 5000.  This will block until the app is listening on that port." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.6773064+03:00" level=info msg="application discovered on port 5000" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.6773064+03:00" level=warning msg="[DEPRECATION NOTICE] Adding a default content type to incoming service invocation requests is deprecated and will be removed in the future. See https://docs.dapr.io/operations/support/support-preview-features/ for more details. You can opt into the new behavior today by setting the configuration option `ServiceInvocation.NoDefaultContentType` to true." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.7713156+03:00" level=info msg="application configuration loaded" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.7720306+03:00" level=warning msg="failed to init actors: no actor state store defined" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.8002235+03:00" level=debug msg="app responded with subscriptions [{messagebus EventTestTopic map[] [0xc0005a09e0] []}]" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.8028561+03:00" level=info msg="app is subscribed to the following topics: [EventTestTopic] through pubsub=messagebus" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.8028561+03:00" level=debug msg="subscribing to topic=EventTestTopic on pubsub=messagebus" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.8054396+03:00" level=debug msg="nats: subscribed to subject EventTestTopic" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T17:59:54.8054396+03:00" level=info msg="dapr initialized. Status: Running. Init Elapsed 3355.6849ms" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:54.8054396+03:00" level=debug msg="Processing NATS Streaming message EventTestTopic/1" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T17:59:55.0639973+03:00" level=warning msg="DROP status returned from app while processing pub/sub event 86ba42de-4874-46d1-8991-c36b78d56a51" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:55.0645752+03:00" level=debug msg="Processing NATS Streaming message EventTestTopic/2" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T17:59:55.2025044+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T17:59:55.2027688+03:00" level=info msg="Error processing operation component[messagebus] input. Retrying..." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:05.2527356+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:05.2529182+03:00" level=error msg="Error processing NATS Streaming message: EventTestTopic/2. Retrying..." app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:10.2671133+03:00" level=debug msg="Processing NATS Streaming message EventTestTopic/2" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:10.2879758+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:10.2879758+03:00" level=info msg="Error processing operation component[messagebus] input. Retrying..." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:20.3201225+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:21.4601797+03:00" level=debug msg="Refreshing all mDNS addresses." app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:21.4601797+03:00" level=debug msg="no mDNS apps to refresh." app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:25.3225501+03:00" level=debug msg="Processing NATS Streaming message EventTestTopic/2" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:25.3437858+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:25.3440371+03:00" level=info msg="Error processing operation component[messagebus] input. Retrying..." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:35.3779068+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:40.3867654+03:00" level=debug msg="Processing NATS Streaming message EventTestTopic/2" app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:40.4129234+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:40.4129234+03:00" level=info msg="Error processing operation component[messagebus] input. Retrying..." app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:50.4535369+03:00" level=warning msg="retriable error returned from app while processing pub/sub event b355dfa2-031b-4bc7-80aa-ef9ef4dc8f1f, topic: EventTestTopic, body: . status code returned: 500" app_id=app instance=PF1SE31X scope=dapr.runtime type=log ver=1.7.0
time="2022-04-13T18:00:51.4729261+03:00" level=debug msg="Refreshing all mDNS addresses." app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
time="2022-04-13T18:00:51.4729261+03:00" level=debug msg="no mDNS apps to refresh." app_id=app instance=PF1SE31X scope=dapr.contrib type=log ver=1.7.0
 ```

**Comment by yaron2** [comment]
@Admiralkheir this is now fixed and will come out in 1.8: https://github.com/dapr/components-contrib/pull/1760.

**Comment by Admiralkheir** [comment]
> @Admiralkheir this is now fixed and will come out in 1.8: #1760.

ty for effort @yaron2
